### PR TITLE
Add SSL context for RedisCluster connections

### PR DESCRIPTION
**Pull request by robin-brabants**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description
This small feature allows to pass an [sslContext](https://www.php.net/manual/en/context.ssl.php) to the [RedisCluster constructor](https://github.com/phpredis/phpredis/blob/develop/cluster.md) from the phpredis extension.

Without this feature, one could not connect to a cluster using TLS. Now when specifying the correct protocol for the seed nodes (e.g.: 'seeds' => ['tls://nodeId-01.xxx:6379', 'tls://nodeId-02.xxx:6379']) and passing an sslContext one can connect to a cluster using TLS .

One can pass an empty array for the sslContext, but also explicitly specify non-default options for the TLS connection. For a full list of ssl context options see: https://www.php.net/manual/en/context.ssl.php.

This for example enables one to connect to an AWS elasticache redis cluster with in-transit encryption set to required due to security reasons.

### CODE REVIEW

**Comment by Ocramius** [comment]
> TARGET THE develop BRANCH????

Is the template still foobar'd ?

File: src/RedisClusterOptions.php
Comment (Ocramius) [inline src/RedisClusterOptions.php]: Is there any way to document this better? :thinking: 

at least `array<string, mixed>`, I suppose? Or `array{verify_peer: bool, ...array<string, mixed>}`?

**Review by Ocramius — COMMENTED** [review]
I understand where this comes from, but the use case of `verify_peer => false` seems kinda anachronistic nowadays.

Possibly better to provision both host and client with a dummy certificate via https://github.com/FiloSottile/mkcert/blob/2a46726cebac0ff4e1f133d90b4e4c42f1edf44a/README.md ?

**Comment by robin-brabants** [comment]
> > TARGET THE develop BRANCH????
> 
> Is the template still foobar'd ?

The template mentions to make a PR for features against a development branch:

Pick the target branch based on the following criteria:
  * New feature, or refactor of existing code: develop branch

File: src/RedisClusterOptions.php
Comment (robin-brabants) [reply src/RedisClusterOptions.php]: Yeah it makes sense to document the field and getter return type as array<string, mixed>.

The single source of truth for recognized options in this array would be: https://www.php.net/manual/en/context.ssl.php. Would it be useful to put this in the docstring?

**Comment by robin-brabants** [comment]
> I understand where this comes from, but the use case of `verify_peer => false` seems kinda anachronistic nowadays.
> 
> Possibly better to provision both host and client with a dummy certificate via https://github.com/FiloSottile/mkcert/blob/2a46726cebac0ff4e1f133d90b4e4c42f1edf44a/README.md ?

it is just an example for a unit test

File: composer.json
Comment (robin-brabants) [inline composer.json:11]: Passing a ssl/tls context to the RedisCluster constructor from the phpredis extension is apparently only introduced in version 5.3.2: https://github.com/phpredis/phpredis/blob/develop/CHANGELOG.md 
I noticed this due to the failing pipeline.
This would of course be a breaking change though.

File: src/RedisClusterOptions.php
Comment (Ocramius) [reply src/RedisClusterOptions.php]: Before throwing effort at it, let's wait for @boesing, who's the main maintainer here :) 

Want to see if we should support the whole SSL context, or just parts of it.

**Comment by robin-brabants** [comment]
I added a few commits which should fix the failing integration tests. Could they be ran again?

File: composer.json
Comment (boesing) [reply composer.json:11]: I havent had this kind of change ever but I remember that I did modified some redis range in the past. IMO having a check during CD that all the required extension versions are fulfilled should be a good thing, so I do not have a problem with that change personally, but how is this kind of stuff handled in other OSS.
Do you have experience here @Ocramius?

File: psalm.xml
Comment (boesing) [inline psalm.xml]: I'd put this into the code via `@psalm-suppress` as that would be reported once its fixed in upstream and thus we are able to remove that suppression.
Suppressions in the psalm config are not reported and thus would suppress this issue for like forever which might introduce problems somewhen in the future.

Would you mind create an issue in psalm as well and/or a PR to update the signature so that we can remove the inline `@psalm-suppression` at some point?
That would be decent value.

File: src/RedisClusterOptions.php
Comment (boesing) [inline src/RedisClusterOptions.php]: do we know what context exists?
Can we probably have an object instead of an array?

File: test/unit/RedisClusterOptionsTest.php
Comment (boesing) [inline test/unit/RedisClusterOptionsTest.php]: I'd rather see an object here.
These are supported SSL context entries: https://www.php.net/manual/en/context.ssl.php

File: test/unit/RedisClusterOptionsTest.php
Comment (boesing) [inline test/unit/RedisClusterOptionsTest.php]: Lets allow instantiation from array shape and object (which needs to be implemented).
Allowing an array with whatever key might introduce problems if keys are passed as `ssl_context` which are not allowed. If we can prevent that from static analysis level, I'd say lets do so.

File: test/unit/RedisClusterOptionsTest.php
Comment (boesing) [inline test/unit/RedisClusterOptionsTest.php]: https://github.com/laminas/laminas-cache-storage-adapter-redis/pull/85/files#r1534542474

**Review by boesing — CHANGES_REQUESTED** [review]
I'd added some comments, would be decent if we can improve this a little bit.
Sorry for the late response, I am not doing too much OSS at the moment.

File: src/RedisClusterOptions.php
Comment (boesing) [reply src/RedisClusterOptions.php]: Oh yah, thats exactly what I requested below.
I'd say lets only return an object, that would allow us to consume that context in a more OOP way. I'd like to prevent arrays as much as possible. IMO we should allow an array shape and an object during instantiation, we could provide a proper Options array shape on class-level, I might want to require that in the next major version anyways for *every* options of an adapter. That would basically allow us to have proper type-safety via static analysis and allows us to have an object instantiated from an array.

Since `laminas-cache` depends on `laminas-stdlib`, we could even implement https://github.com/laminas/laminas-stdlib/blob/ec5f3d3f1a2c44575d3369981ff5a1058aacf77d/src/ArraySerializableInterface.php to have a method to pass the SSL context as an array to the `RedisCluster#__construct`.

WDYT? 

----

I stumbled over `template-type` annotation in phpstan and raised a feature request (which I might implement on my own at some point) in psalm: https://github.com/vimeo/psalm/issues/10849

**Comment by boesing** [comment]
I wonder if we can add an integration test where we actually setup a redis cluster with SSL and self-signed certificate to verify if this is actually working. I don't expect this in this PR tho as I think it would already suffice and I can understand that taking time to implement this kind of setup would be too much.
It took me quite some time to even have redis cluster docker setup up and running, since the docker image we are using here does not support SSL right now, we would have to create an own docker image. Thus I think I will create an issue off this comment once we merged this PR.

**Comment by robin-brabants** [comment]
Implemented an sslContext class right now which contains the [sslcontext](https://www.php.net/manual/en/context.ssl.php) options in the fields. 

I omitted the 'capture_peer_cert' and 'capture_peer_cert_chain' options since they are not relevant for us in my opinion.

I wanted to make some of the default values clearly visible in the constructor. I made them correspond to the actual sslContext default options at least for more recent php versions with the exception of 'SNI_enabled' which would make sense in my opinion to enable by default if possible.

File: psalm.xml
Comment (robin-brabants) [reply psalm.xml]: Makes sense, I changed this. Also made a PR for changing this in psalm and referenced it in the docstring: https://github.com/vimeo/psalm/pull/10862

**Comment by robin-brabants** [comment]
@Ocramius @boesing 
Are there still any changes required for this PR

**Comment by boesing** [comment]
Sorry, havent reviewed yet. Will probably do so either today or tomorrow evening.

**Comment by boesing** [comment]
I refactored the code a little bit, especially since I am not a fan from those abbreviations all round tech.
I renamed some properties but still preserve the `context` notation when converting the object to the `ssl context` array.

I'd say we should only allow 1:1 array notations which would be accepted by PHP as well, thats why I removed all that camel case handling and just have two dedicated methods, a named factory to instantiate our object from an SSL context array (in the notation supported by PHP) and a method which converts the object back to a PHP supported SSL context array.

I also do not allow `RedisClusterOptions#setSslContext` to consume an array. IMHO if that is being called from somewhere, it should be either `null` or the `SslContext` object.
I did override `AdapterOptions#setFromArray` to ensure that `ssl_context` or `sslContext` array keys are always converted to an object in case an array is passed.
So test is still passing.

I also removed all the ssl context defaults. That prevents us from having issues with several PHP versions where (whyever) defaults might change and then we have to see how to provide defaults for both versions. I do not really see that we should do that and thus I'd say lets allow passing `null` for any option and in case it is not explicitly set (i.e. `null`), we do not pass that in the `array` we pass to `RedisCluster#__construct`.

I also would say we should not require `ext-openssl` since it would be only required from upstream projects as this component does pretty much work without that extension (as I removed all the defaults which were using `OPENSSL_*` constants) and thus I'd say, that is a requirement which should be added by projects rather than this library.

----

I'd love to get some review of my changes and some feedback. If that works for you and you still see the feature is serving your needs, I'd love to get a quick 👍🏼  so that we can proceed. 
Since I do not have a cluster serving via TLS, I am unable to verify if everything is working. I've created #88 to keep track that we should have integration tests up and running so that we do not accidentally break stuff.

But I'd be fine with releasing this without actual integration tests as its a new feature which needs to be explicitly implemented by projects and thus its okayish to depend on projects having integration testings up and running until we implement #88.

Thanks for your massive work here and I'm happy to see your feedback.

----

_Had to rebase with latest 2.8.x as renovate updated `composer.lock`..._

**Review by boesing — APPROVED** [review]

**Comment by robin-brabants** [comment]
Sorry that I only see your improvement right now, but was a bit preoccupied.
I would have some time again after the weekend and will give my feedback on Monday.

File: src/SslContext.php
Comment (robin-brabants) [inline src/SslContext.php:7]: nice addition, but it is only supported from php 8.2 onwards

File: test/unit/SslContextTest.php
Comment (robin-brabants) [inline test/unit/SslContextTest.php:1]: maybe it is overkill, but we could include some testcase for the  'toSslContextArray' instance method if we provide some non-null parameters to the constructor

**Comment by robin-brabants** [comment]
@boesing 
Your changes look great to me! I just had two minor comments.
For me it is fine to merge this PR as is.

File: src/SslContext.php
Comment (boesing) [reply src/SslContext.php:7]: Yah, it has no impact on PHP 8.1 but there is not an issue either (unless some1 parses attributes and tries to instantiate them). I think this is okay but still good find.

File: test/unit/SslContextTest.php
Comment (boesing) [reply test/unit/SslContextTest.php:1]: Its not explicitly tested but implicitly via the array notations.
I'd be fine with that for now, that does already verify if expected properties are set with expected values and that would be good enough for me.
Lets ship, we can still add more QA stuff in future releases but for now we still have 100% coverage.

**Comment by boesing** [comment]
Thanks @robin-brabants!

**Pull request merged**
